src/retrace/retrace.py

```diff
@@ -477,66 +472,92 @@ def is_package_known(package_nvr, arch, releaseid=None):
 
     return any([os.path.isfile(f) for f in candidates])
 
-# tricky
-# crash is not able to process the vmcore from different arch
-# (not even x86_64 and x86). In addition, there are several
-# types of vmcores: running 'file' command on el5- vmcore results
-# into an expected description (x86-64 or 80386 coredump), while
-# el6+ vmcores are just proclaimed 'data'. Another thing is that
-# the OSRELEASE in the vmcore sometimes contains architecture
-# and sometimes it does not.
+
+#
+# In real-world testing, approximately 60% of the time the kernel
+# version of a vmcore can be identified with the crash tool.
+# In the other 40% of the time, we must use some other method.
+#
+# The below function contains a couple regex searches that work
+# across a wide variety of vmcore formats and kernel versions.
+# We do not attempt to identify the file type since this is often not
+# reliable, but we assume the version information exists in some form
+# in the first portion of the file.  Testing has indicated that we do
+# not need to scan the entire file but can rely on a small portion
+# at the start of the file, which helps preserve useful pages in the
+# OS page cache.
+#
+# The following regex's are used for the 40% scenario
+# 1. Look for 'OSRELEASE='.  For example:
+# OSRELEASE=2.6.18-406.el5
+# NOTE: We can get "OSRELEASE=%" so we disallow the '%' character after the '='
+OSRELEASE_VAR_PARSER = re.compile("OSRELEASE=([^%][^\x00\s\n\r]+)")
```

Comment:
[^\x00\s\n\r] means - "does not contain NULL, whitespace-characters, new line and return". However \s is equivalent to [ \t\n\r\f\v] so the "\n\r" is redundant. Also note that there is \S which is [^ \s]. Only reason why not using \S is the \x00. Why is there this NULL character? I cannot quite imagine where this could pop up.

Reply:
Null is needed as a null terminator of the string.  I originally did not have null but had to add it to ensure the regex used with the open/read method was equivalent to the regex used with the 'strings' method.  Remember the goal is to remove 'strings' and so I'm just trying to not lose any detection while improving it.  I actually was skeptical myself until I saw one vmcore and opened it to find that the kernel version was terminated by a NULL (I think it was a RHEL5 version).  If you think this is not clear, I may be able to break this into a "matches" regex and a "non-matches" regex:
 +OSRELEASE_VAR_PARSER = re.compile("OSRELEASE=([^%][\S]+)[\x00\s]+")

I have not tested the above and so it would need another round of testing,  but I think that would work.  Is that clearer? If you think it's clearer I am ok with doing another round of testing.

Reply:
> until I saw one vmcore and opened it to find that the kernel version was terminated by a NULL

Ok, when it is required that it is fine to have

When we need to have the `\x00` then I would not go into `\S` only would drop `\n\r`, so the result would be
`OSRELEASE_VAR_PARSER = re.compile("OSRELEASE=([^%][^\x00\s]+)")`

Reply:
ok yeah that sounds good